# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def add_two(x):
    return x + 2

print(add_two(3))

5


In [3]:
print(add_two)

<function add_two at 0x7fddd06617e0>


In [4]:
# lambda param: one-line function
lambda x: x + 2

<function __main__.<lambda>(x)>

In [5]:
# IIFE - Immediately Invoked Function Expression
# Execute a lambda function immediately after defining the function
# Syntax: (lambda param: expression)(arg)
(lambda x: x + 2)(3)

5

#### Saving to a Variable

In [6]:
f = lambda x,y:x+y+2
print(f)

<function <lambda> at 0x7fddccd1c430>


In [7]:
f(3,2)

7

#### Multiple Inputs

In [9]:
# Syntax: lambda param1, param2, etc: expression
l = 3
w = 5
h = 8

(lambda x,y,z: x*y*z)(l,w,h)


120

In [ ]:
# As an IIFE - (lambda param1, param2, etc: expression)(arg1, arg2, etc.)


#### Passing a Function into a Function

In [10]:
def apply_func(func, arg):
    return func(arg)

In [12]:
def my_math(x):
    return x*x


In [16]:
apply_func(my_math,5)

25

#### Passing a Lambda into a Function

In [ ]:
# Call the apply_to_each function, pass in a lambda function as f argument


In [18]:
apply_func(lambda x: x*x, 5)

25

#### Returning a Function from a Function

In [28]:
def create_enclosure():
    num = 0
    def inner_func():
        nonlocal num
        num=num+1
        return num
    return inner_func

my_multiplier = create_enclosure()
print(my_multiplier())
print(my_multiplier())
print(my_multiplier())
print(my_multiplier())
print(my_multiplier())

1
2
3
4
5


In [29]:
def create_multiplier(num):
    def inner_func(x):
        return num*x
    return inner_func
my_multiplier = create_multiplier(2)
my_multiplier(4)

8

In [30]:
my_multiplier(8)

16

#### Returning a Lambda from a Function

In [31]:
def create_multiplier(num):
    return lambda x: num*x

my_multiplier = create_multiplier(2)
my_multiplier(4)

8

#### If Statements within Lambdas

In [32]:
def do_something(num):
    if num>10:
        return num * 2
    else:
        return num + 2
    
do_something(5)

7

In [ ]:
# Ternary: value_if_true if condition else value_if_false


In [34]:
def do_something(num):
    return num *2 if num > 10 else num + 2
do_something(5)

7

In [35]:
(lambda num: num *2 if num > 10 else num + 2)(5)

7

In [40]:
# write a lambda function that takes a name and and age and returns
# a string that says  ""[name] is [this many] years old"

(lambda name, age: f"{name} is {age} years old")("Peter", 40)

'Peter is 40 years old'

In [41]:
# write a lambda function that takes a 3 numbers and returns those numbers added together
(lambda n1,n2,n3: n1+n2+n3)(10,20,30)

60

In [49]:
student = {"name":"Random Person", "IQ": randint(1,240)}
print(student["IQ"])

40


In [46]:
from random import randint

my_class = [
    {"name":"Valencia", "IQ": randint(1,240)},
    {"name":"Mason", "IQ": randint(1,240)},
    {"name":"Cynthia", "IQ": randint(1,240)},
    {"name":"Sheridan", "IQ": randint(1,240)}
]

print(my_class)

[{'name': 'Valencia', 'IQ': 120}, {'name': 'Mason', 'IQ': 64}, {'name': 'Cynthia', 'IQ': 157}, {'name': 'Sheridan', 'IQ': 80}]


In [50]:
sorted(my_class, key=lambda student: student["IQ"])

[{'name': 'Mason', 'IQ': 64},
 {'name': 'Sheridan', 'IQ': 80},
 {'name': 'Valencia', 'IQ': 120},
 {'name': 'Cynthia', 'IQ': 157}]

In [51]:
sorted(my_class, key=lambda student: student["IQ"], reverse=True)

[{'name': 'Cynthia', 'IQ': 157},
 {'name': 'Valencia', 'IQ': 120},
 {'name': 'Sheridan', 'IQ': 80},
 {'name': 'Mason', 'IQ': 64}]

In [42]:
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [58]:
print(my_class)

[{'name': 'Valencia', 'IQ': 120}, {'name': 'Mason', 'IQ': 64}, {'name': 'Cynthia', 'IQ': 157}, {'name': 'Sheridan', 'IQ': 80}]


In [61]:
my_class.sort(key=lambda student: student["IQ"], reverse=True)
print(my_class)

[{'name': 'Cynthia', 'IQ': 157}, {'name': 'Valencia', 'IQ': 120}, {'name': 'Sheridan', 'IQ': 80}, {'name': 'Mason', 'IQ': 64}]


In [57]:
help(list.sort)

Help on method_descriptor:

sort(self, /, *, key=None, reverse=False)
    Sort the list in ascending order and return None.
    
    The sort is in-place (i.e. the list itself is modified) and stable (i.e. the
    order of two equal elements is maintained).
    
    If a key function is given, apply it once to each list item and sort them,
    ascending or descending, according to their function values.
    
    The reverse flag can be set to sort in descending order.



## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

In [62]:
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [66]:
def squared(num1, num2):
    if num1 < 10 and num2 < 10:
        return (num1**2, num2**2)
    else:
        return (num1, num2)
    
nums1 = [4,10,3,1,6]
nums2 = [4,1,20,3,15]

new_list =[]
for i in range(len(nums1)):
    new_list.append(squared(nums1[i], nums2[i]))
    
print(new_list)


[(16, 16), (10, 1), (3, 20), (1, 9), (6, 15)]


In [68]:
# map(function, *iterables -> List, Tuple, String, Set, Dictionary, etc)
squared_nums_map = map(squared, nums1, nums2)
print(list(squared_nums_map))

[(16, 16), (10, 1), (3, 20), (1, 9), (6, 15)]


In [72]:
def squared(num1, num2):
    return (num1**2, num2**2) if num1 < 10 and num2 < 10 else (num1, num2)
         
    
         
squared(4,4)    

(16, 16)

#### Using Lambda's with Map

In [74]:
squared_nums_map = map((lambda num1, num2: (num1**2, num2**2) if num1 < 10 and num2 < 10 else (num1, num2)), nums1, nums2)
print(list(squared_nums_map))

[(16, 16), (10, 1), (3, 20), (1, 9), (6, 15)]


#### In-Class Exercise #1 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [77]:
nums = [4,2,6,8,15]
# [7, 3, 11, 15, 29]
list(map(lambda num: num * 2 - 1, nums))


[7, 3, 11, 15, 29]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

In [78]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [85]:
bool({"key":123})

True

In [88]:
# Syntax: filter(function_to_apply, iterable -> List,Tuple,Set,Dict,String,etc)
list(filter(None, [True, True, False, "", "string", [], [1,2,4,5] ]))

[True, True, 'string', [1, 2, 4, 5]]

True

#### Using Lambda's with Filter()

In [101]:
names = ["Daniel", "Leisa", "Sheridan", "Desiree", "Keanu", "Santa", "Danny Devito"]
def d_names(name):
    if name[0].lower() == 'd':
        return True
    else:
        return False
    
the_d_names = filter(d_names, names)
list(the_d_names)

['Daniel', 'Desiree', 'Danny Devito']

In [102]:
lamb_ds = filter(lambda name: True if name[0].lower() == 'd' else False, names)
list(lamb_ds)


['Daniel', 'Desiree', 'Danny Devito']

#### In-Class Exercise #2 <br>
<p>Using the filter function, create a list out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [107]:
36 <= 24

False

In [103]:
nums = [2,7,4.2,1.6,9,4.4,4.9]

In [108]:
from statistics import mean
below_mean = filter(lambda x: x <= mean(nums) ,nums)
list(below_mean)


[2, 4.2, 1.6, 4.4]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [111]:
from functools import reduce

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, iterable[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence
    or iterable, from left to right, so as to reduce the iterable to a single
    value.  For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the iterable in the calculation, and serves as a default when the
    iterable is empty.



In [117]:
list_1 = [2,4,6,8,10]

def subtract_nums(num1, num2):
    return num1 - num2
result = reduce(subtract_nums, list_1)
print(result)

-26


In [112]:
2-4-6-8-10

-26

#### Using Lambda's with Reduce()

In [118]:
the_sum = reduce(lambda aggregator, current_value: aggregator + current_value ,list_1)
the_sum

30

#### In-Class Exercise #3 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [120]:
result = reduce((lambda total, next_num: total*next_num), list_1)
result

3840

In [123]:
test = "x" + "y"
test

'xy'

In [125]:
letters = ["C", "o", "m", "e", " ", "o", "n", " ", "D", "o","w","n"]
full_phrase = reduce(lambda phrase, letter: phrase + letter, letters)
full_phrase


'Come on Down'

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [139]:
def recursive(i):
    if i == 4: #STOP LOOPING -- Base Case
        return
    print("hi!")
    i+=1
    recursive(i)
    

In [140]:
recursive(0)

hi!
hi!
hi!
hi!


#### Implementing a Base Case

#### Writing a Factorial Function

In [143]:
# 5! = 5 * 4!
# 4! = 4 * 3!
# ...
# 1! = 1

def factorial(num):
    if num <= 1:
        print('factorial(1)=1')
        return 1
    else:
        print(f"factorial({num})={num} * factorial({num-1})")
        return num * factorial(num-1)
factorial(5000)
        

factorial(50000)=50000 * factorial(49999)
factorial(49999)=49999 * factorial(49998)
factorial(49998)=49998 * factorial(49997)
factorial(49997)=49997 * factorial(49996)
factorial(49996)=49996 * factorial(49995)
factorial(49995)=49995 * factorial(49994)
factorial(49994)=49994 * factorial(49993)
factorial(49993)=49993 * factorial(49992)
factorial(49992)=49992 * factorial(49991)
factorial(49991)=49991 * factorial(49990)
factorial(49990)=49990 * factorial(49989)
factorial(49989)=49989 * factorial(49988)
factorial(49988)=49988 * factorial(49987)
factorial(49987)=49987 * factorial(49986)
factorial(49986)=49986 * factorial(49985)
factorial(49985)=49985 * factorial(49984)
factorial(49984)=49984 * factorial(49983)
factorial(49983)=49983 * factorial(49982)
factorial(49982)=49982 * factorial(49981)
factorial(49981)=49981 * factorial(49980)
factorial(49980)=49980 * factorial(49979)
factorial(49979)=49979 * factorial(49978)
factorial(49978)=49978 * factorial(49977)
factorial(49977)=49977 * factorial

factorial(47368)=47368 * factorial(47367)
factorial(47367)=47367 * factorial(47366)
factorial(47366)=47366 * factorial(47365)
factorial(47365)=47365 * factorial(47364)
factorial(47364)=47364 * factorial(47363)
factorial(47363)=47363 * factorial(47362)
factorial(47362)=47362 * factorial(47361)
factorial(47361)=47361 * factorial(47360)
factorial(47360)=47360 * factorial(47359)
factorial(47359)=47359 * factorial(47358)
factorial(47358)=47358 * factorial(47357)
factorial(47357)=47357 * factorial(47356)
factorial(47356)=47356 * factorial(47355)
factorial(47355)=47355 * factorial(47354)
factorial(47354)=47354 * factorial(47353)
factorial(47353)=47353 * factorial(47352)
factorial(47352)=47352 * factorial(47351)
factorial(47351)=47351 * factorial(47350)
factorial(47350)=47350 * factorial(47349)
factorial(47349)=47349 * factorial(47348)
factorial(47348)=47348 * factorial(47347)
factorial(47347)=47347 * factorial(47346)
factorial(47346)=47346 * factorial(47345)
factorial(47345)=47345 * factorial

RecursionError: maximum recursion depth exceeded while calling a Python object

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [ ]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York", "DC"]


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [ ]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]



### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]
`

In [ ]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>
<code>fib(1) = 0
fib(2) = 1
fib(3) = 1
fib(4) = 2
fib(5) = 3
fib(6) = 5
fib(7) = 8
fib(8) = 13
fib(9) = 21
fib(10) = 34
fib(11) = 55
fib(12) = 89
fib(13) = 144
</code>

### Exercise #5<br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [ ]:
def my_gen(number):
    pass

for num in my_gen(10):
    print(num)
    
# 100
# 81
# 64
# 49
# 36
# 25
# 16
# 9
# 4
# 2
# 1
# 0